histogram/prometheus_model.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package histogram
+
+import (
+	"fmt"
+
+	dto "github.com/prometheus/client_model/go"
+	model "github.com/prometheus/prometheus/model/histogram"
+)
+
+type APIBucket[BC model.BucketCount] struct {
+	Boundaries   uint64
+	Lower, Upper float64
+	Count        BC
+}
+
+func NewModelHistogram(ch *dto.Histogram) (*model.Histogram, *model.FloatHistogram) {
+	if ch.GetSampleCountFloat() > 0 || ch.GetZeroCountFloat() > 0 {
+		// It is a float histogram.
+		fh := model.FloatHistogram{
+			Count:           ch.GetSampleCountFloat(),
+			Sum:             ch.GetSampleSum(),
+			ZeroThreshold:   ch.GetZeroThreshold(),
+			ZeroCount:       ch.GetZeroCountFloat(),
+			Schema:          ch.GetSchema(),
+			PositiveSpans:   make([]model.Span, len(ch.GetPositiveSpan())),
+			PositiveBuckets: ch.GetPositiveCount(),
+			NegativeSpans:   make([]model.Span, len(ch.GetNegativeSpan())),
+			NegativeBuckets: ch.GetNegativeCount(),
+		}
+		for i, span := range ch.GetPositiveSpan() {
+			fh.PositiveSpans[i].Offset = span.GetOffset()
+			fh.PositiveSpans[i].Length = span.GetLength()
+		}
+		for i, span := range ch.GetNegativeSpan() {
+			fh.NegativeSpans[i].Offset = span.GetOffset()
+			fh.NegativeSpans[i].Length = span.GetLength()
+		}
+		return nil, &fh
+	}
+	h := model.Histogram{
+		Count:           ch.GetSampleCount(),
+		Sum:             ch.GetSampleSum(),
+		ZeroThreshold:   ch.GetZeroThreshold(),
+		ZeroCount:       ch.GetZeroCount(),
+		Schema:          ch.GetSchema(),
+		PositiveSpans:   make([]model.Span, len(ch.GetPositiveSpan())),
+		PositiveBuckets: ch.GetPositiveDelta(),
+		NegativeSpans:   make([]model.Span, len(ch.GetNegativeSpan())),
+		NegativeBuckets: ch.GetNegativeDelta(),
+	}
+	for i, span := range ch.GetPositiveSpan() {
+		h.PositiveSpans[i].Offset = span.GetOffset()
+		h.PositiveSpans[i].Length = span.GetLength()
+	}
+	for i, span := range ch.GetNegativeSpan() {
+		h.NegativeSpans[i].Offset = span.GetOffset()
+		h.NegativeSpans[i].Length = span.GetLength()
+	}
+	return &h, nil
+}
+
+func BucketsAsJson[BC model.BucketCount](buckets []APIBucket[BC]) [][]interface{} {
+	ret := make([][]interface{}, len(buckets))
+	for i, b := range buckets {
+		ret[i] = []interface{}{b.Boundaries, fmt.Sprintf("%v", b.Lower), fmt.Sprintf("%v", b.Upper), fmt.Sprintf("%v", b.Count)}
+	}
+	return ret
+}
+
+func GetAPIBuckets(h *model.Histogram) []APIBucket[uint64] {
```

Comment:
This and `GetAPIFloatBuckets` are exported function since I intend to use them for the web ui changes.